src/ModelGradient.cpp

```diff
@@ -428,5 +430,468 @@ void MFSparseGradient::check_values() const {
+
+    } else {
+      // 2 -> dense structure
+      start_time_ttemp = get_time_ms();
+      store_value<int16_t>(mem, dense_type);
+      for (int j = 0; j < K; ++j) {
+        store_value<int16_t>(
+            mem, change_nvt_ptr->operator[](slice_map[word_idx] * K + j));
+      }
+      time_ttemp += (get_time_ms() - start_time_ttemp) / 1000.0;
+    }
+  }
+
+  time_nvt_find += (get_time_ms() - start_time_temp) / 1000.0;
+
+  for (int i = 0; i < K; ++i) {
+    store_value<int32_t>(mem, change_nt_ptr->operator[](i));
+  }
+
+  for (int i = 0; i < len; ++i) {
+    store_value<int32_t>(mem, fixed_slices[slice_id][i]);
+  }
+
+  start_time_temp = get_time_ms();
+
+  uncompressed_size = sizeof(int32_t) * (2 + K + len) +
+                      sizeof(int16_t) * (len + S + 2 * N + (len - S) * K);
+  size_t max_compressed_size = LZ4_compressBound(uncompressed_size);
+  char* compressed_mem = new char[uncompressed_size];
+  time_temp += (get_time_ms() - start_time_temp) / 1000.0;
+
+  to_send_size = LZ4_compress_default(mem_begin, compressed_mem,
+                                      uncompressed_size, max_compressed_size);
+  time_compress += (get_time_ms() - start_time_temp) / 1000.0;
+
+  double time_taken = (get_time_ms() - start_time_func) / 1000.0;
+
+  delete mem_begin;
+  time_whole += (get_time_ms() - start_time_benchmark) / 1000.0;
+  return compressed_mem;
+}
+
+int LDAUpdates::pre_assign_slices(int slice_size) {
+  int num_slices = slice.size() / slice_size;
+
+  // mapping global word id to its slice id
+  std::array<int, 1000000> gindex_2_slice_id;
+  gindex_2_slice_id.fill(-1);
+
+  std::cout << "Global vocab dim: " << slice.size()
+            << " slice_size: " << slice_size << " num_slices: " << num_slices
+            << std::endl;
+
+  // memory reserved for pre-cached word indices
+  fixed_slices.clear();
+  fixed_slices.reserve(num_slices);
+  for (int i = 0; i < num_slices; ++i) {
+    std::vector<int> to_push;
+    to_push.reserve(slice_size);
+    fixed_slices.push_back(to_push);
+  }
+
+  // sequentially assign word id to slice
+  int cur = 0;
+  for (int i = 0; i < slice.size(); ++i) {
+    fixed_slices[cur].push_back(slice[i]);
+    gindex_2_slice_id[slice[i]] = cur;
+    cur += 1;  // sequentially
+    if (cur == num_slices) {
+      cur = 0;
+    }
+  }
+
+  w_slices.clear();
+  w_slices.reserve(ws_ptr->size());
+  for (int i = 0; i < ws_ptr->size(); ++i) {
+    std::vector<std::vector<int>> w_slice_i;
+    w_slice_i.resize(num_slices, std::vector<int>());
+    for (int j = 0; j < num_slices; ++j) {
+      w_slice_i[j].reserve(slice.size());
+    }
+    for (int j = 0; j < ws_ptr->operator[](i).size(); ++j) {
+      int gindex = ws_ptr->operator[](i)[j];
+      // w_slice_i[p] := word indices for the p^{th} vocab slice
+      w_slice_i[gindex_2_slice_id[gindex]].push_back(j);
+    }
+    w_slices.push_back(w_slice_i);
+  }
+
+  ws_ptr.reset();
+
+  global_v = slice.size();
```

Comment:
Can we get rid of this variable and use slice.size() in get_partial_model() -- the only other place where this is used?

Reply:
Done.